src/v/serde/avro/tests/parser_test.cc

```diff
@@ -0,0 +1,547 @@
+                if (!result) {
+                    return AssertionFailure() << fmt::format(
+                             "array element mismatch - {}", result.message());
+                }
+            }
+            return AssertionSuccess();
+        }
+        case avro::AVRO_MAP: {
+            auto expected_map = expected.value<::avro::GenericMap>();
+            auto actual_map = actual.value<::avro::GenericMap>();
+            if (expected_map.value().size() != actual_map.value().size()) {
+                return AssertionFailure() << fmt::format(
+                         "map size mismatch {} != {}",
+                         expected_map.value().size(),
+                         actual_map.value().size());
+            }
+            for (size_t i : boost::irange(expected_map.value().size())) {
+                if (
+                  expected_map.value()[i].first
+                  != actual_map.value()[i].first) {
+                    return AssertionFailure() << fmt::format(
+                             "map key mismatch {} != {}",
+                             expected_map.value()[i].first,
+                             actual_map.value()[i].first);
+                }
+                auto result = are_equal(
+                  expected_map.value()[i].second, actual_map.value()[i].second);
+                if (!result) {
+                    return AssertionFailure() << fmt::format(
+                             "map value mismatch - {}", result.message());
+                }
+            }
+            return AssertionSuccess();
+        }
+
+        case avro::AVRO_UNION:
+            if (expected.unionBranch() != actual.unionBranch()) {
+                return AssertionFailure() << fmt::format(
+                         "union branch mismatch {} != {}",
+                         expected.unionBranch(),
+                         actual.unionBranch());
+            }
+            return are_equal(
+              expected.value<::avro::GenericUnion>().datum(),
+              actual.value<::avro::GenericUnion>().datum());
+        case avro::AVRO_FIXED: {
+            auto expected_fixed = expected.value<::avro::GenericFixed>();
+            auto actual_fixed = expected.value<::avro::GenericFixed>();
+            if (expected_fixed.value() != actual_fixed.value()) {
+                return AssertionFailure()
+                       << fmt::format("fixed value mismatch");
+            }
+            return AssertionSuccess();
+        }
+        case avro::AVRO_UNKNOWN:
+        case avro::AVRO_NUM_TYPES:
+            break;
+        }
+        return AssertionFailure();
+    }
+
+    iobuf serialize_with_avro(
+      const ::avro::GenericDatum& datum, const ::avro::ValidSchema& schema) {
+        std::unique_ptr<::avro::OutputStream> out
+          = ::avro::memoryOutputStream();
+        auto e = ::avro::binaryEncoder();
+        e->init(*out);
+        avro_write(e, schema, datum);
```

Comment:
nit: if you prefer there's also avro_iobuf_ostream in iceberg/avro_utils.h that you can pull out